crates/ursa-gateway/src/cache/tlrfu.rs

```diff
@@ -0,0 +1,469 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    sync::Arc,
+    time::UNIX_EPOCH,
+};
+
+use anyhow::{bail, Context, Result};
+use util::timer::now;
+
+use super::lru::Lru;
+
+struct Data {
+    value: Vec<u8>,
+    freq: usize,
+    lru_k: usize,
+    _tll: u128,
+}
+
+pub struct Tlrfu {
+    store: HashMap<Arc<String>, Data>,
+    freq: BTreeMap<usize, Lru<usize, Arc<String>>>, // shrinkable
+    ttl: BTreeMap<u128, Arc<String>>,
+    used_size: u64,
+    _max_size: u64,
+    _ttl_buf: u128,
+}
+
+impl Tlrfu {
+    pub fn new(max_size: u64, ttl_buf: u128) -> Self {
+        Self {
+            store: HashMap::new(),
+            freq: BTreeMap::new(),
+            ttl: BTreeMap::new(),
+            used_size: 0,
+            _max_size: max_size,
+            _ttl_buf: ttl_buf,
+        }
+    }
+
+    fn _contains(&self, k: &String) -> bool {
+        self.store.contains_key(k)
+    }
+
+    fn _is_size_exceeded(&self, bytes: u64) -> bool {
+        self.used_size + bytes > self._max_size
+    }
+
+    pub async fn get(&mut self, k: &String) -> Result<Option<&Vec<u8>>> {
+        if let Some(data) = self.store.get_mut(k) {
+            let lru = self
+                .freq
+                .get_mut(&data.freq)
+                .with_context(|| format!("[TLRFU]: Key: {k} not found at freq {}.", data.freq))?;
+            let key = lru.remove(&data.lru_k).await.with_context(|| {
+                format!(
+                    "[TLRFU]: Failed to remove LRU key: {} not found at freq {}.",
+                    data.lru_k, data.freq
+                )
+            })?;
+            lru.is_empty().then(|| self.freq.remove(&data.freq));
+            data.freq += 1;
```

Comment:
We're assuming that this is very unlikely to overflow?

Reply:
Yes, buckets are unlikely to overflow.